scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -488,6 +488,55 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
     }
   }
 
+  /**
+   * Return an SCollection with the pairs from `this` whose keys are in `that`
+   * when the cardinality of `this` >> `that`, but neither can fit in memory
+   * (see [[PairHashSCollectionFunctions.hashIntersectByKey]]).
+   * @group per key
+   */
+  def sparseIntersectByKey(that: SCollection[K], thatNumKeys: Int, fpProb: Double = 0.1)(
+    implicit koder: Coder[K],
+    voder: Coder[V],
+    hash: Hash128[K]): SCollection[(K, V)] = {
+    val bfSettings =
+      PairSCollectionFunctions.optimalBFSettings(thatNumKeys, fpProb)
+    if (bfSettings.numBFs == 1) {
+      sparseIntersectByKeyImpl(that, thatNumKeys.toInt, fpProb)
+    } else {
+      val n = bfSettings.numBFs
+      val thisParts = self.partition(n, _._1.hashCode() % n)
+      val thatParts = that.partition(n, _.hashCode() % n)
+      val joined = (thisParts zip thatParts).map {
+        case (lhs, rhs) =>
+          lhs.sparseIntersectByKeyImpl(rhs, bfSettings.capacity, fpProb)
+      }
+      SCollection.unionAll(joined)
+    }
+  }
+
+  protected def sparseIntersectByKeyImpl(that: SCollection[K],
+                                         thatNumKeys: Int,
+                                         fpProb: Double = 0.1)(
+    implicit koder: Coder[K],
+    voder: Coder[V],
+    hash: Hash128[K]): SCollection[(K, V)] = {
+    val width = BloomFilter.optimalWidth(thatNumKeys, fpProb).get
+    val numHashes = BloomFilter.optimalNumHashes(thatNumKeys, width)
+    val rhsBf = that
+      .aggregate(BloomFilterAggregator[K](numHashes, width))
+      .asIterableSideInput
+
+    self
+      .withSideInputs(rhsBf)
+      .filter { case (e, c) => c(rhsBf).headOption.exists(_.maybeContains(e._1)) }
+      .toSCollection
+      .cogroup(that.map((_, ())))
```

Comment:
We don't need the `cogroup+flatMap` anymore since it's covered by `filter`?

Reply:
i guess it depends on if we expect this function to be approximate or exact -- since the bloom filter can have false positives

Reply:
Maybe add a `exact: Boolean` option that defaults to false? If the user joins the result with another data set it defeats the purpose of `cogroup+flatMap`.